File: packages/engine-ts/src/narrative-display-text.ts
import type { Model, SessionState } from './types.js'
import { expandTemplate, expandTemplateWithTracking } from './template.js'
import { findMatchingTemplates } from './conversation-templates.js'
import type { DescriptionState } from './description-tracker.js'

/**
 * SP-TGEN-001: レガシーな `{flag:…}` / `{resource:…}` 等を先に展開する（段階0）。
 * Web Tester 由来の互換用。`expandTemplate` の `{name}` 形式と併存する。
 */
export function applyLegacySessionPlaceholders(text: string, session: SessionState): string {
  let resolved = text

  Object.entries(session.flags ?? {}).forEach(([key, value]) => {
    resolved = resolved.replace(new RegExp(`\\{flag:${escapeRegExp(key)}\\}`, 'g'), value ? 'true' : 'false')
  })

  Object.entries(session.resources ?? {}).forEach(([key, value]) => {
    resolved = resolved.replace(new RegExp(`\\{resource:${escapeRegExp(key)}\\}`, 'g'), String(value))
  })

  Object.entries(session.variables ?? {}).forEach(([key, value]) => {
    resolved = resolved.replace(new RegExp(`\\{variable:${escapeRegExp(key)}\\}`, 'g'), String(value))
  })

  resolved = resolved.replace(/\{nodeId\}/g, session.nodeId)
  resolved = resolved.replace(/\{time\}/g, String(session.time))

  return resolved
}

function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

export interface ResolveNarrativeDisplayTextOptions {
  /**
   * false のとき ConversationTemplate の末尾連結を行わない（GUI プレビュー等）。
   * 既定は true（プレイ本線と同一）。
   */
  appendConversationTemplates?: boolean
}

export interface ResolveNarrativeDisplayTextTrackedOptions extends ResolveNarrativeDisplayTextOptions {
  /** SP-TGEN 段階2: 描写追跡の入力状態（省略時は空） */
  descriptionState?: DescriptionState
  /** `[entity~]` 選択の決定論シード（既定 0） */
  descriptionSeed?: number
}

export interface ResolveNarrativeDisplayTextTrackedResult {
  text: string
  descriptionState: DescriptionState
}

/**
 * プレイヤー向けにノード本文を決定論的に合成する（SP-TGEN-001 段階0〜3、追跡なし）。
 *
 * 順序: レガシー置換 → expandTemplate（entities/events がある場合のみ）→ 会話テンプレート末尾連結（任意）。
 * `[entity~]` 描写追跡が必要な場合は `resolveNarrativeDisplayTextTracked` を使う。
 */
export function resolveNarrativeDisplayText(
  rawText: string,
  model: Model,
  session: SessionState,
  options?: ResolveNarrativeDisplayTextOptions
): string {
  if (!rawText || !session) return rawText

  const appendTemplates = options?.appendConversationTemplates !== false

  let resolved = applyLegacySessionPlaceholders(rawText, session)

  if (model?.entities || session?.events) {
    resolved = expandTemplate(resolved, model, session)
  }

  if (
    appendTemplates &&
    model?.conversationTemplates &&
    model.conversationTemplates.length > 0 &&
    session?.events
  ) {
    const matches = findMatchingTemplates(model.conversationTemplates, session, model)
    if (matches.length > 0) {
      const insertions = matches.map(m => m.expandedText).join(' ')
      resolved = `${resolved} ${insertions}`
    }
  }

  return resolved
}

/**
 * 段階0〜3に加え SP-TEXT の `[entity~]` 描写追跡（段階2）を含む合成。
 * 段階1は `expandTemplateWithTracking` で一括する（`[~]` が無い場合も `expandTemplate` と同結果）。
 */
export function resolveNarrativeDisplayTextTracked(
  rawText: string,
  model: Model,
  session: SessionState,
  options?: ResolveNarrativeDisplayTextTrackedOptions
): ResolveNarrativeDisplayTextTrackedResult {
  if (!rawText || !session) {
    return { text: rawText, descriptionState: options?.descriptionState ?? {} }
  }

  const appendTemplates = options?.appendConversationTemplates !== false
  const descIn = options?.descriptionState ?? {}
  const seed = options?.descriptionSeed ?? 0

  let resolved = applyLegacySessionPlaceholders(rawText, session)
  let descriptionState = descIn

  if (model?.entities || session?.events) {
    const tracked = expandTemplateWithTracking(resolved, model, session, descIn, seed)
    resolved = tracked.text
    descriptionState = tracked.descriptionState
  }

  if (
    appendTemplates &&
    model?.conversationTemplates &&
    model.conversationTemplates.length > 0 &&
    session?.events
  ) {
    const matches = findMatchingTemplates(model.conversationTemplates, session, model)
    if (matches.length > 0) {
      const insertions = matches.map(m => m.expandedText).join(' ')
      resolved = `${resolved} ${insertions}`
    }
  }

  return { text: resolved, descriptionState }
}

File: packages/sdk-unity/Runtime/NarrativeDisplayText.cs
using System;
using System.Collections.Generic;
using System.Linq;
using NarrativeGen.Runtime;

namespace NarrativeGen
{
    /// <summary>
    /// engine-ts の <c>resolveNarrativeDisplayTextTracked</c> に相当（段階0〜3 + 会話テンプレ連結）。
    /// </summary>
    public sealed class ResolveNarrativeDisplayTextTrackedOptions
    {
        public IReadOnlyDictionary<string, HashSet<string>>? DescriptionState { get; set; }
        public int DescriptionSeed { get; set; }
        public bool AppendConversationTemplates { get; set; } = true;
        public IReadOnlyDictionary<string, int>? TemplateUsageState { get; set; }
    }

    public sealed class ResolveNarrativeDisplayTextTrackedResult
    {
        public string Text { get; set; } = string.Empty;
        public Dictionary<string, HashSet<string>> DescriptionState { get; set; } =
            new(StringComparer.OrdinalIgnoreCase);
        public Dictionary<string, int> TemplateUsageState { get; set; } =
            new(StringComparer.OrdinalIgnoreCase);
    }

    public static class NarrativeDisplayText
    {
        public static ResolveNarrativeDisplayTextTrackedResult ResolveNarrativeDisplayTextTracked(
            string? rawText,
            NarrativeModel model,
            Session session,
            ResolveNarrativeDisplayTextTrackedOptions? options = null)
        {
            var emptyState = CloneDescriptionState(options?.DescriptionState);
            if (string.IsNullOrEmpty(rawText))
            {
                return new ResolveNarrativeDisplayTextTrackedResult
                {
                    Text = rawText ?? string.Empty,
                    DescriptionState = emptyState
                };
            }

            var appendTemplates = options?.AppendConversationTemplates ?? true;
            var descIn = options?.DescriptionState;
            var seed = options?.DescriptionSeed ?? 0;
            var usageState = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
            if (options?.TemplateUsageState != null)
            {
                foreach (var kv in options.TemplateUsageState)
                    usageState[kv.Key] = kv.Value;
            }

            string resolved;
            Dictionary<string, HashSet<string>> descriptionState;

            var body = rawText ?? string.Empty;
            var hasEntities = model.Entities != null && model.Entities.Count > 0;
            if (hasEntities || session.Events.Count > 0)
            {
                var tracked = NarrativeText.ExpandTemplateWithTracking(body, model, session, descIn, seed);
                resolved = tracked.Text;
                descriptionState = tracked.DescriptionState;
            }
            else
            {
                resolved = NarrativeText.ApplyLegacyPlaceholders(body, session);
                descriptionState = CloneDescriptionState(descIn);
            }

            if (appendTemplates &&
                model.ConversationTemplates is { Count: > 0 } templates &&
                session.Events.Count > 0)
            {
                var matches = ConversationTemplateMatcher.FindMatchingTemplates(templates, session, model, usageState);
                if (matches.Count > 0)
                {
                    var insertions = string.Join(" ", matches.Select(m => m.ExpandedText));
                    resolved = string.IsNullOrEmpty(resolved) ? insertions : resolved + " " + insertions;
                    foreach (var match in matches)
                    {
                        usageState = ConversationTemplateMatcher.RecordTemplateUsage(usageState, match.TemplateId);
                    }
                }
            }

            return new ResolveNarrativeDisplayTextTrackedResult
            {
                Text = resolved,
                DescriptionState = descriptionState,
                TemplateUsageState = usageState
            };
        }

        private static Dictionary<string, HashSet<string>> CloneDescriptionState(
            IReadOnlyDictionary<string, HashSet<string>>? src)
        {
            var d = new Dictionary<string, HashSet<string>>(StringComparer.OrdinalIgnoreCase);
            if (src == null) return d;
            foreach (var kv in src)
                d[kv.Key] = new HashSet<string>(kv.Value, StringComparer.OrdinalIgnoreCase);
            return d;
        }
    }
}
